README.md
# MatryoshkaView

Maintains a list of inner (subset/nested) views and their geometic boundaries for a particular table. Helps you spawn new inner views and lookup the right one.

## Installation

Add this line to your application's Gemfile:

```ruby
gem 'matryoshka_view'
```

And then execute:

    $ bundle

Or install it yourself as:

    $ gem install matryoshka_view

## Usage

1. create a MatryoshkaView object with the base table name (`MatryoshkaView.new(base: Place)`)
2. maybe try to lookup a view that satisfies a boundary (`MatryoshkaView.new(base: Place).lookup(the_geom_geojson_of_massachusetts)`)
3. maybe create a new view within a boundary (`MatryoshkaView.new(base: Place).spawn(the_geom_geojson_of_boston)`)

## Contributing

1. Fork it ( https://github.com/[my-github-username]/matryoshka_view/fork )
2. Create your feature branch (`git checkout -b my-new-feature`)
3. Commit your changes (`git commit -am 'Add some feature'`)
4. Push to the branch (`git push origin my-new-feature`)
5. Create a new Pull Request


CHANGELOG
2015-06-08 / 0.2.6

* Bug fixes

  * Restrict lookup to the same base

2015-05-12 / 0.2.5

* Enhancements

  * Use geom_source.the_geom if geom_source has changed

2014-10-19 / 0.2.4

* Enhancements

  * Faster - no need for ST_Expand with &&

2014-10-12 / 0.2.3

* Enhancements

  * Speed things up by just using && for bbox check; ST_Contains not needed.

2014-10-10 / 0.2.2

* Enhancements

  * Speed things up with an explicit bbox check followed by the buffer query

2014-10-09 / 0.2.1

* Bug fixes

  * Properly determine containers - before we were inadvertently drawing a box around container geoms, which resulted in false positives

2014-08-21 / 0.2.0

* Breaking changes

  * Switch to CREATE TABLE ... LIKE instead of materialized views

2014-08-20 / 0.1.2

* Bug fixes

  * Don't ignore custom name when you find an old view

* Enhancements

  * Provide MatryoshkaView::Record.cleanup

2014-08-20 / 0.1.1

* Enhancements

  * One weird trick to make ST_Contains slightly less strict (so that stuff on the border of a view counts as contained)

2014-08-20 / 0.1.0

* Breaking changes

  * MatryoshkaView#lookup now takes geom_source: or the_geom_geojson: arguments

* Bug fixes

  * Fix use of ST_Contains during view creation

0.0.1 / 2014-08-20

initial release


spec/spec_helper.rb
$LOAD_PATH.unshift File.expand_path('../../lib', __FILE__)

require 'pry'

require 'logger'
require 'fileutils'

log_dir = File.expand_path('../../log', __FILE__)
FileUtils.mkdir_p log_dir

logger = Logger.new(File.join(log_dir, 'test.log'))
logger.level = Logger::DEBUG

dbname = 'matryoshka_view_test'
unless ENV['FAST'] == 'true'
  system 'dropdb', '--if-exists', dbname
  system 'createdb', dbname
  system 'psql', dbname, '--command', 'CREATE EXTENSION postgis'
  system 'psql', dbname, '--command', 'CREATE TABLE places (id serial primary key, the_geom geometry(Geometry,4326), the_geom_webmercator geometry(Geometry,3857))'
end

require 'active_record'

ActiveRecord::Base.logger = logger

ActiveRecord::Base.establish_connection "postgresql://127.0.0.1/#{dbname}"

# http://gray.fm/2013/09/17/unknown-oid-with-rails-and-postgresql/
ActiveRecord::ConnectionAdapters::PostgreSQLAdapter.tap do |klass|
  klass::OID.register_type('geometry', klass::OID::Identity.new)
end

require 'rspec'
require 'database_cleaner'
require 'the_geom_geojson/examples'

DatabaseCleaner.strategy = :deletion, {except: %w{ spatial_ref_sys }}

RSpec.configure do |config|
  config.fail_fast = true

  config.before(:suite) do
    ActiveRecord::Base.connection.execute "DROP SCHEMA IF EXISTS #{MatryoshkaView::SCHEMA_NAME} CASCADE"
  end

  config.before(:each) do
    ActiveRecord::Base.connection.execute "DROP SCHEMA IF EXISTS #{MatryoshkaView::SCHEMA_NAME} CASCADE"
    DatabaseCleaner.start
  end

  config.after(:each) do
    DatabaseCleaner.clean
  end
end

require 'matryoshka_view'

MatryoshkaView.setup

class Place < ActiveRecord::Base
  include TheGeomGeoJSON::ActiveRecord
end


spec/matryoshka_view_spec.rb
require 'spec_helper'

describe MatryoshkaView do
  let(:world)               { MatryoshkaView.new(base: Place.table_name) }

  def geojson(name)
    TheGeomGeoJSON::EXAMPLES.fetch name
  end

  def place(name)
    place = Place.new
    place.save!
    place.the_geom_geojson = geojson(name)
    place.save!
    place
  end

  shared_examples 'OK' do
    it "just gives you back the base by default" do
      expect(world.from_sql).to eq('places')
    end

    it "doesn't auto-create anything" do
      expect(world.lookup(the_geom_geojson: geojson(:burlington_point))).to eq(world)
      expect(world.lookup(the_geom_geojson: geojson(:montpelier_point))).to eq(world)
    end

    it "helps you spawn inner views given geojson" do
      burlington # spawn it
      expect(MatryoshkaView.view_exists?(burlington.from_sql)).to be_truthy
    end

    it "lets you specify an name name" do
      world.spawn the_geom_geojson: geojson(:burlington), name: 'magic'
      expect(MatryoshkaView.view_exists?('magic')).to be_truthy
      ActiveRecord::Base.connection.execute 'DROP TABLE magic CASCADE'
    end

    describe "after spawning a matryoshka view" do
      before do
        burlington
      end

      it "tells you what view to use inside boundaries (inclusive)" do
        expect(world.lookup(the_geom_geojson: geojson(:burlington_point))).to eq(burlington)
      end

      it "falls back to original table outside boundaries" do
        expect(world.lookup(the_geom_geojson: geojson(:montreal_canada))).to eq(world)
      end

      xit "doesn't confuse bases" do
        # you can only use this with one base table
      end
    end

    describe "non-overlapping matryoshka views" do
      before do
        south_burlington
        burlington_downtown
      end

      it "chooses the right view" do
        expect(world.lookup(the_geom_geojson: geojson(:south_burlington_point))).to eq(south_burlington)
        expect(world.lookup(the_geom_geojson: geojson(:burlington_downtown_point))).to eq(burlington_downtown)
      end

      it "falls back to original table outside boundaries" do
        expect(world.lookup(the_geom_geojson: geojson(:montreal_canada))).to eq(world)
      end
    end

    describe "overlapping matryoshka views" do
      before do
        burlington
        south_burlington
      end

      it "chooses the smaller view" do
        expect(world.lookup(the_geom_geojson: geojson(:south_burlington_point))).to eq(south_burlington)
      end
    end

    describe "contents of matryoshka views" do
      it "has the same columns" do
        expect(ActiveRecord::Base.connection.columns(burlington.from_sql).map(&:name)).to match_array(ActiveRecord::Base.connection.columns(world.from_sql).map(&:name))
      end
    end

    describe "optimizing view creation" do
      # even if you tell a view it's based on nationwide, it should see if there is a smaller view that it can base itself on
    end
  end

  describe 'with the_geom_geojson' do
    let(:burlington)          { world.spawn the_geom_geojson: geojson(:burlington) }
    let(:south_burlington)    { world.spawn the_geom_geojson: geojson(:south_burlington) }
    let(:burlington_downtown) { world.spawn the_geom_geojson: geojson(:burlington_downtown) }

    it_behaves_like 'OK'
  end

  describe 'with geom_source' do
    let(:burlington)          { world.spawn geom_source: place(:burlington) }
    let(:south_burlington)    { world.spawn geom_source: place(:south_burlington) }
    let(:burlington_downtown) { world.spawn geom_source: place(:burlington_downtown) }

    it_behaves_like 'OK'
  end

  describe 'with the_geom' do
    let(:burlington)          { world.spawn geom_source: place(:burlington) }
    it "tells you what view to use inside boundaries (inclusive)" do
      burlington
      expect(world.lookup(geom_source: place(:burlington_point))).to eq(burlington)
    end

    it "falls back to original table outside boundaries" do
      burlington
      expect(world.lookup(geom_source: place(:montreal_canada))).to eq(world)
    end

    it "doesn't confuse bases" do
      other_world = MatryoshkaView.new(base: 'gibberish')
      burlington
      expect(other_world.lookup(geom_source: place(:burlington_point))).to eq(other_world)
    end
  end

end


lib/matryoshka_view.rb
require 'active_record'
require 'hash_digest'

require 'active_support'
require 'active_support/core_ext'

require 'matryoshka_view/version'
require 'matryoshka_view/record'

class MatryoshkaView
  class << self
    def setup
      Record.create_table
    end

    # @private
    def logger
      ActiveRecord::Base.logger
    end

    def view_exists?(*expected_names)
      result = ActiveRecord::Base.connection_pool.with_connection do |conn|
        conn.select_values <<-SQL
          SELECT 1
          FROM   pg_catalog.pg_class mv
          WHERE
            mv.oid::regclass::text IN (#{expected_names.map { |name| conn.quote(name) }.join(',')})
            AND mv.relkind = 'r'
        SQL
      end
      result.length == expected_names.length
    end
  end

  SCHEMA_NAME = 'matryoshka_view'

  attr_reader :base
  attr_reader :geom_source

  def initialize(base:, geom_source: nil, the_geom_geojson: nil, name: nil)
    @base = base
    @the_geom_geojson = the_geom_geojson
    @geom_source = geom_source
    @name = name
  end

  def lookup(geom_source: nil, the_geom_geojson: nil)
    # FIXME move to Record class method
    hit = with_connection do |c|
      if geom_source
        if geom_source.changed?
          Record.where("(base = #{c.quote base}) AND (the_geom && #{c.quote geom_source.the_geom}) AND ST_Contains(ST_Buffer(the_geom, 0.00001), #{c.quote geom_source.the_geom})").order("ST_Area(the_geom, false) ASC").first
        else
          Record.where("(base = #{c.quote base}) AND (the_geom && (SELECT the_geom FROM #{geom_source.class.quoted_table_name} WHERE id = #{c.quote(geom_source.id)})) AND ST_Contains(ST_Buffer(the_geom, 0.00001), (SELECT the_geom FROM #{geom_source.class.quoted_table_name} WHERE id = #{c.quote(geom_source.id)}))").order("ST_Area(the_geom, false) ASC").first
        end
      elsif the_geom_geojson
        Record.where("(base = #{c.quote base}) AND (the_geom && ST_SetSRID(ST_GeomFromGeoJSON(#{c.quote the_geom_geojson}), 4326)) AND ST_Contains(ST_Buffer(the_geom, 0.00001), ST_SetSRID(ST_GeomFromGeoJSON(#{c.quote the_geom_geojson}), 4326))").order("ST_Area(the_geom, false) ASC").first
      else
        raise "expecting geom_source or the_geom_geojson"
      end
    end
    hit.try(:view) || self
  end

  def spawn(attrs)
    child = MatryoshkaView.new attrs.reverse_merge(base: base)
    child.spawn!
    child
  end

  def spawn!
    with_connection do |c|
      c.execute "CREATE SCHEMA IF NOT EXISTS #{SCHEMA_NAME}"
      c.execute <<-SQL
        CREATE TABLE #{name} (LIKE #{quoted_base} INCLUDING ALL)
      SQL
      c.execute <<-SQL
        INSERT INTO #{name}
          SELECT *
          FROM #{quoted_base}
          WHERE ST_Contains(ST_SetSRID(ST_GeomFromGeoJSON(#{c.quote(the_geom_geojson)}), 4326), #{quoted_base}.the_geom)
      SQL
      record = Record.new
      record.name = name
      record.base = base
      record.save!
      record.the_geom_geojson = the_geom_geojson
      record.save!
    end
  end

  def the_geom_geojson
    @the_geom_geojson ||= geom_source.try :the_geom_geojson
  end

  def quoted_base
    @quoted_base ||= with_connection do |c|
      base.split('.', 2).map { |part| c.quote_column_name(part) }.join('.')
    end
  end

  def name
    @name ||= if inner?
      "#{SCHEMA_NAME}.t#{HashDigest.digest3(uniq_attrs)}"
    else
      base
    end
  end

  def from_sql
    name
  end

  def name_without_schema
    @name_without_schema ||= name.split('.', 2).last
  end

  def eql?(other)
    # puts "#{uniq_attrs} <-> #{other.uniq_attrs}"
    other.is_a?(MatryoshkaView) and uniq_attrs == other.uniq_attrs
  end
  alias == eql?

  def uniq_attrs
    {
      base_table_name: base,
      the_geom_geojson: the_geom_geojson,
    }
  end

  private

  def inner?
    !the_geom_geojson.nil?
  end

  # db helper to make sure we immediately return connections to pool
  # @private
  [
    # :execute,
    :quote,
    # :quote_column_name,
    # :quote_table_name,
    # :select_all,
    # :select_rows,
    # :select_value,
    # :select_values,
  ].each do |method_id|
    define_method method_id do |sql|
      with_connection do |conn|
        conn.send method_id, sql
      end
    end
  end

  # db helper to make sure we immediately return connections to pool
  # @private
  def with_connection
    ActiveRecord::Base.connection_pool.with_connection do |conn|
      yield conn
    end
  end

end


lib/matryoshka_view/record.rb
require 'the_geom_geojson'
require 'the_geom_geojson/active_record'

class MatryoshkaView
  class Record < ActiveRecord::Base
    class << self
      def create_table
        activity = false
        connection_pool.with_connection do |c|
          unless table_exists?
            c.execute "CREATE TABLE #{quoted_table_name}(name text UNIQUE NOT NULL)"
            activity = true
          end
          existing_columns = column_names
          if (missing_columns = COLUMNS.keys - column_names).any?
            activity = true
            add_columns = missing_columns.map {|name| %{ADD COLUMN "#{name}" #{COLUMNS[name]}} }
            c.execute "ALTER TABLE #{quoted_table_name} #{add_columns.join(',')}"
            missing_columns.each do |name|
              case name
              when /the_geom/
                c.execute "CREATE INDEX ON #{quoted_table_name} USING gist(#{name})"
              else
                c.execute "CREATE INDEX ON #{quoted_table_name} (#{name})"
              end
            end
          end
          if activity
            c.schema_cache.clear!
            reset_column_information
          end
        end
      end

      def cleanup
        find_each do |record|
          record.destroy unless record.exists?
        end
      end
    end

    include TheGeomGeoJSON::ActiveRecord

    self.table_name = 'matryoshka_view_records'
    self.primary_key = 'name'

    COLUMNS = {
      'base'  => 'text',
      'the_geom' => 'geometry(Geometry,4326)',
      'the_geom_webmercator' => 'geometry(Geometry,3857)',
    }

    def view
      @view ||= begin
        save!
        unless exists?
          raise "missing #{name} (#{inspect})"
        end
        MatryoshkaView.new name: name, base: base, the_geom_geojson: the_geom_geojson
      end
    end

    def exists?
      MatryoshkaView.view_exists? name
    end

  end
end
